### DISCUSSION TITLE
Make formal CSS syntax more understandable

### DISCUSSION THREAD

**Discussion by SebastianZ** [mdn/mdn-community#207] [Content]
Previously, there were already discussions about what to do with the formal syntax shown on CSS pages.

Because the syntax isn't intuitive, it is obviously something that confuses readers as well as contributors as @hamishwillee noted in https://github.com/mdn/content/issues/2202#issuecomment-799020728.

With the rise of CSS Houdini and its [CSS Properties and Values API](https://www.w3.org/TR/css-properties-values-api-1/#syntax-strings) to define custom properties the formal syntax becomes more important, though. Authors need to understand it to some extent in order to be able to define the syntax for their custom properties.

Therefore, I believe we should find a way to make the syntax more understandable. At some point we [displayed some info icon when hovering the syntax box](https://groups.google.com/g/mozilla.dev.mdc/c/ucaUxwA2xIQ/m/UJ8L_ptXAQAJ). Somehow linking to an explanatory page like the [value definition syntax page](https://developer.mozilla.org/en-US/docs/Web/CSS/Value_definition_syntax) would surely help people getting to know about it.

Besides that, the different parts of the syntax are actually already linking to that page. That feature is just broken at the moment, see https://github.com/mdn/yari/issues/2963.

Sebastian

**Comment by wbamberg** [comment]
Hi Sebastian :)

I'm not sure exactly what you're suggesting here - we do already show info on hover and link to the syntax documentation in places like https://developer.mozilla.org/en-US/docs/Web/CSS/mask#formal_syntax don't we (I mean, given a fix for https://github.com/mdn/yari/issues/2963)? Or are you asking about something else?

About a hundred years ago I looked into pretty-printing the formal syntax: https://discourse.mozilla.org/t/pretty-printing-the-css-formal-syntax-part-2/27330. I thought it was quite nice but there wasn't enough interest to make it ship. I think there's also a lot of potential to use css-tree to make formal syntax more interactive, so you could explore a bit of syntax and drill down into its components. It seems like it would be a significant amount of work to spec and implement though. There would also be a question about how much of this should be done in a macro rather than in the platform directly.

**Comment by SebastianZ** [comment]
Hi Will! Great to read from you again! 😃

Basically, I just wanted to (re)start a discussion here on how to improve its display as I am hearing for years now that it's hard to understand for readers.

Thank you for sharing the Discourse link! I see now that you already put much thought into it. And I totally agree with all what you wrote there and here.

I guess with "not enough interest" you mean that it wasn't a priority because everyone in the thread seemed to be very positive about the suggested changes and your prototype.

Just to sum up the points from that thread here:

* Break the syntax into its constituent parts
* Break function parameters
* Initially collapse it and allow to expand it (i.e. use summary/details)
* Use [CSSTree](https://github.com/csstree/csstree) for pretty-printing and syntax highlighting
* Keep the links and (translated) tooltips to basic data types, combinators, and multipliers
* Maybe add a legend for combinators and multipliers

Please let me know if I missed something!

I don't know what the final decision about macros was. I thought you wanted to get rid of them completely at some point. Though if they are going to stay, the `{{csssyntax}}` macro could be extended to do most of this.

Sebastian

**Comment by SebastianZ** [comment]
> With the rise of CSS Houdini and its CSS Properties and Values API to define custom properties the formal syntax becomes more important, though. Authors need to understand it to some extent in order to be able to define the syntax for their custom properties.

To clarify this, CSS Properties and Values allows authors to define custom properties using the same formal syntax for their values as for the predefined properties.

The specification also [provides some examples](https://www.w3.org/TR/css-properties-values-api-1/#examples):

```css
@property --radius {
  syntax: "<length>";
  inherits: false;
  initial-value: 0px;
}
```

Where the `syntax` descriptor can take any kind of formal syntax.

The spec. also has this example for a syntax definition:

```
"foo | <color># | <integer>"
```

@hamishwillee That's what I meant in #2202.

Sebastian

**Comment by wbamberg** [comment]
> I guess with "not enough interest" you mean that it wasn't a priority because everyone in the thread seemed to be very positive about the suggested changes and your prototype.

Basically people didn't (don't?) think formal syntax is important enough to spend much time improving, and/or just wanted to remove it or hide it in some way. I have some sympathy with that, it definitely only serves a niche among our users, so if you user-test it 95% of people will just say "this is baffling", and it's then hard to prioritise. But I can also see the argument that serving the 5% is important as well. Also I think we both agree that we could do a much better job of presenting it. So it's a bit chicken and egg: it looks not very useful, which makes it hard to invest time into making it more useful :).

> Just to sum up the points from that thread here:
> 
> * Break the syntax into its constituent parts
> * Break function parameters
> *  Initially collapse it and allow to expand it (i.e. use summary/details)
> * Use CSSTree for pretty-printing and syntax highlighting
> * Keep the links and (translated) tooltips to basic data types, combinators, and multipliers
> * Maybe add a legend for combinators and multipliers

Yeah, something like that :). As past-me said in that post, I tried using legends and I think the problem is that when the syntax is big, the legend is a long way away from the syntax, if you see what I mean. But, legends are more accessible, so maybe worth more thought.

The prototype code is at https://github.com/wbamberg/kumascript/blob/better-formal-syntax/macros/csssyntax.ejs. I guess it might be challenging convincing people to allow css-tree as a new dependency in the platform, but I do think it's worth using css-tree if we can, it would make life much easier.

> I don't know what the final decision about macros was. I thought you wanted to get rid of them completely at some point. Though if they are going to stay, the {{csssyntax}} macro could be extended to do most of this.

I think eventually something like csssyntax could become platform code (as is the code to render compat tables today actually) but that's a long way off.

**Comment by Rumyra** [comment]
Hey @wbamberg - how resolved is this discussion with the modifications to csssyntax?

**Reply by wbamberg** [reply]
@SebastianZ , what do you think? I don't have any more immediate plans/ideas for improving formal syntax.

**Comment by SebastianZ** [comment]
I'd say, let's have a look at [`<color>`](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#formal_syntax) and check the points mentioned earlier in this thread. 😃
Disclaimer: I am probably not the best person to judge the display because I read it for many years now and got used to it over the years.

>     * Break the syntax into its constituent parts

Check.
 
>     * Break function parameters

We don't have that at the moment. That's ok for me, though again, I am not the usual audience.

>     * Initially collapse it and allow to expand it (i.e. use summary/details)

This is one point that should still be considered. @wbamberg You initially mentioned that 95% of users actually don't care about it. So I think it's not much of a burden to ask the remaining 5% (including me) to expand it.

>     * Use CSSTree for pretty-printing and syntax highlighting

There is obviously some syntax highlighting and pretty-printing now. Did CSSTree make it into the code, in the end?
With the breaks and the basic highlightin the syntax is already much more readable. Some small tweaks could still improve the readability, I guess, like displaying function names the same way as keywords. As mentioned above, there could be a way to break function parameters into their own lines, which might make them more readable for some people, but I don't know.

>     * Keep the links and (translated) tooltips to basic data types, combinators, and multipliers

Those are there. What I am missing though, and what we had earlier, are the links to the different data types. E.g. `<colorspace-params>` should have an anchor link to its definition.
Also, general data types which are used in many different places like `<length-percentage>` should probably get their own page like `<length>` and `<percentage>` have it instead of showing them as entry in the syntax. This would shorten the syntax block a little. See [`grid-template-columns`](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-columns) for example.

>     * Maybe add a legend for combinators and multipliers

Will's comment to that:

> I tried using legends and I think the problem is that when the syntax is big, the legend is a long way away from the syntax, if you see what I mean. But, legends are more accessible, so maybe worth more thought.

We still don't have a legend for them but we at least link to the [syntax definition page](https://developer.mozilla.org/en-US/docs/Web/CSS/Value_definition_syntax).

Though I still believe a proper legend would make the syntax a little easier to understand.
I imagine a little (i) icon or the word "Legend" at the top right corner of the syntax box. It would be sticky positioned so it's always visible. Clicking on it would show the legend explaining the different combinators and multipliers and also link to the syntax definition page.

-----

To summarize it, the formal syntax is already more readable than it used to be, though there's still room for improvement.

Sebastian

**Reply by wbamberg** [reply]
Hi Sebastian! Thanks for the reply.

> > ```
> > * Break function parameters
> > ```
> 
> We don't have that at the moment. That's ok for me, though again, I am not the usual audience.

There are more pretty-printing things we could consider but I'm not proposing to add any myself for now. What's there at the moment is simple and tries to strike a balance between being useful in most cases and keeping the syntax short. I think it would be easier now to extend it than before, but like I say I don't have any immediate plans to do this.

> > ```
> > * Initially collapse it and allow to expand it (i.e. use summary/details)
> > ```
> 
> This is one point that should still be considered. @wbamberg You initially mentioned that 95% of users actually don't care about it. So I think it's not much of a burden to ask the remaining 5% (including me) to expand it.

Yeah, I changed my mind about this :). I special-cased `<color>` and `<gradient>`, so instead of expanding them I link out to the separate page (e.g. https://developer.mozilla.org/en-US/docs/Web/CSS/border#formal_syntax) and I think with that the formal syntax is in general not too long.

> 
> > ```
> > * Use CSSTree for pretty-printing and syntax highlighting
> > ```
> 
> There is obviously some syntax highlighting and pretty-printing now. Did CSSTree make it into the code, in the end? With the breaks and the basic highlightin the syntax is already much more readable. Some small tweaks could still improve the readability, I guess, like displaying function names the same way as keywords. As mentioned above, there could be a way to break function parameters into their own lines, which might make them more readable for some people, but I don't know.

Yes, it uses CSSTree now, which makes it a lot easier to do things like this, and extend it to do different highlighting etc if we want to.

> > ```
> > * Keep the links and (translated) tooltips to basic data types, combinators, and multipliers
> > ```
> 
> Those are there. What I am missing though, and what we had earlier, are the links to the different data types. E.g. `<colorspace-params>` should have an anchor link to its definition. Also, general data types which are used in many different places like `<length-percentage>` should probably get their own page like `<length>` and `<percentage>` have it instead of showing them as entry in the syntax. This would shorten the syntax block a little. See [`grid-template-columns`](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-columns) for example.

So for internal links, like in `<colorspace-params>`, I'll have a think about that.

For general data types like `<length-percentage>`, I suppose that's really asking for an extension of [the existing thing where we link out to `<color>` and `<gradient>`](https://github.com/mdn/yari/blob/25219ce47801d40ff42286fb84bcfb6b8592c309/kumascript/macros/CSSSyntax.ejs#L19-L20)? I think this is tricky though - see https://github.com/orgs/mdn/discussions/138#discussioncomment-3035090 for more on this. Basically I think there's a balance between keeping the syntax short and avoiding too much fragmentation. If people always have to go chasing across lots of pages to get a complete picture, this doesn't seem ideal.

> > ```
> > * Maybe add a legend for combinators and multipliers
> > ```
> 
> Will's comment to that:
> 
> > I tried using legends and I think the problem is that when the syntax is big, the legend is a long way away from the syntax, if you see what I mean. But, legends are more accessible, so maybe worth more thought.
> 
> We still don't have a legend for them but we at least link to the [syntax definition page](https://developer.mozilla.org/en-US/docs/Web/CSS/Value_definition_syntax).
> 
> Though I still believe a proper legend would make the syntax a little easier to understand. I imagine a little (i) icon or the word "Legend" at the top right corner of the syntax box. It would be sticky positioned so it's always visible. Clicking on it would show the legend explaining the different combinators and multipliers and also link to the syntax definition page.

I think this is worth exploring but it's not a thing I'm likely to do as it has a design dependency. I will file an issue though in case people want to pursue it.

**Reply by wbamberg** [reply]
-> https://github.com/mdn/yari/issues/6962

**Comment by dontcallmedom** [comment]
I think we could automatically annotate declarations in the Formal Syntax with comments that give a human-readable interpretation of the line.

For instance, when `border` is declared as `  <line-width>  ||  <line-style>  ||  <color>`, automatically annotating this as comment above the declaration as "`border` is composed of a `<line-width>` value, a `<line-style>` value and a `<color>` value, in any order" would help interpretation and also improve the ability of readers to make sense of the syntax over time.

Not all syntaxes would lead to easily human readable interpretations, so maybe this would only be done for relative simple syntaxes (for some definition of that :), but I assume this would already cover quite a bit of them.

(writing such an interpreter isn't trivial, but nor should it be a major piece of work)

**Reply by wbamberg** [reply]
That's an interesting idea! We'd have to have a think about how hard it would be - or rather, how often it would get too hard :). I like the idea that we could help people learn the syntax in this way.
